zerver/lib/actions.py

```diff
@@ -3927,39 +3930,36 @@ def do_update_user_presence(user_profile: UserProfile,
                             status: int) -> None:
     client = consolidate_client(client)
 
+    # TODO: While we probably DO want creating an account to
+    # automatically create a first `UserPresence` object with
+    # last_connected_time and last_active_time as the current time,
+    # our presence tests don't understand this, and it'd be perhaps
+    # wrong for some cases of account creation via the API.  So we may
+    # want a "never" value here as the default.
     defaults = dict(
-        timestamp=log_time,
-        status=status,
+        last_active_time=log_time,
+        last_connected_time=log_time,
         realm_id=user_profile.realm_id
     )
 
     (presence, created) = UserPresence.objects.get_or_create(
         user_profile = user_profile,
-        client = client,
         defaults = defaults
     )
 
-    stale_status = (log_time - presence.timestamp) > datetime.timedelta(minutes=1, seconds=10)
-    was_idle = presence.status == UserPresence.IDLE
-    became_online = (status == UserPresence.ACTIVE) and (stale_status or was_idle)
-
-    # If an object was created, it has already been saved.
-    #
-    # We suppress changes from ACTIVE to IDLE before stale_status is reached;
-    # this protects us from the user having two clients open: one active, the
-    # other idle. Without this check, we would constantly toggle their status
-    # between the two states.
-    if not created and stale_status or was_idle or status == presence.status:
-        # The following block attempts to only update the "status"
-        # field in the event that it actually changed.  This is
-        # important to avoid flushing the UserPresence cache when the
-        # data it would return to a client hasn't actually changed
-        # (see the UserPresence post_save hook for details).
-        presence.timestamp = log_time
-        update_fields = ["timestamp"]
-        if presence.status != status:
-            presence.status = status
-            update_fields.append("status")
+    now_online = (log_time - presence.last_active_time) > datetime.timedelta(minutes=5)
```

Comment:
Arguably, this threshhold should be that 70s constant described above, aka at least a minute less than OFFLINE_THRESHOLD_SECS and at least 10s more than the presence polling frequency.

But I think it might be rational for it to just be `OFFLINE_THRESHOLD_SECS - 10` .... maybe let's set it to that for now.

Reply:
makes sense, changed